File: api/apitest.rb
#!/usr/bin/ruby
require 'rubygems'
require 'rest_client'
require 'json'
require 'pp'
require 'hexdump'

DEBUG = 1

class EnterpriseServer
  attr_reader :token,:expiration

  def initialize(h = {})
    @host = h[:host]
    @port = 8443

    f = open('.apipass')
    userinfo = JSON.parse(f.read)[@host]
    @user = userinfo['user']
    @pass = userinfo['pass']
    @base_url = "https://#{@host}:#{@port}/api/mt"

    if (h[:token])
      @token = h[:token]
    else
      resp = self.getSessionToken :username=> @user, :password => @pass
      @token = resp['sessionToken']
      @expiration = resp['expiration'].to_i / 1000
    end
  end

  def method_missing(m,*args)
    json = (args.size == 0)? {} : args[0]
    if @token
      json['sessionToken'] = @token
    end

    data = CGI::escape(json.to_json)
    puts "POST #{@base_url}/#{m}" if DEBUG
    jj json if DEBUG
    RestClient.post("#{@base_url}/#{m}?","json=#{data}") do |response, request, result, &block|
      case response.code
      when 200..299
        resp = JSON.parse(URI::decode(response))
        jj resp if DEBUG
        return resp
      else
        response.return!(request,result,&block)
      end
    end
  end

  def print_expiration
    puts Time.at(es.expiration).to_s
  end

  def print_all_devices
    (self.searchForEnterprise :enterprise => {})['enterprises'].each do |ent|
      (self.getAllAcquirements :enterprise => {:domain => ent['domain']})['users'].each do |user|
        user['devices'].each do |device|
          p = device['provisioned']? '*' : ' '
          printf "%-16s%-32s #{p} %-18s %-18s\n", ent['name'], user['email'],device['mac'], device['imei']
        end
      end
    end
  end

  def print_device(user)
    (self.searchForUser :user => {:email => user})['users'].each do |user|
      user['devices'].each do |dev|
        (self.searchForDevice :device => {:id => dev['id']})['device']['settings'].sort {|a,b| a['key'] <=> b['key']}.each do |s|
          if s['key'] == 'wifi_provision_list'
            puts s['value']
            s['value'].hexdump
          end
          
#          printf "%-32s%-16s%s\n", s['key'],s['value'],s['complianceState']
        end
      end
    end
  end

  def print_all_policies
    (self.searchForEnterprise :enterprise => {})['enterprises'].each do |ent|
      (self.searchForPolicy :enterprise => {:domain => ent['domain']})['policies'].each do |policy|
        puts "#{policy['name']}"
        # (self.getPolicy :policy => {:id => policy['id']})['settings'].each do |setting|
        #   printf "    %-16s:%s\n", setting['key'],setting['value']
        # end
      end
    end
  end

  def update_policy(dom,pol)
    (self.searchForPolicy :enterprise => {:domain => dom}, :policy => {:searchTerm => pol})['policies'].each do |policy|
      puts "#{policy['name']}"
      self.updatePolicy :policy => {
        :id => policy['id'],
        :new_settings => [{ :key => 'apn_provision_list',
                            :value => "{'apnArr':[{'apn':'123','authtype':'NONE','bearer':'UNSPECIFIED','mcc':'123','mmsc':'mmsc','mmsport':'123','mmsproxy':'mmsproxy','mnc':'123','name':'foo','password':'password','port':'123','protocol':'IPV6','proxy':'proxy','roaming_protocol':'IP','server':'server','type':'type','user':'user'}]}"}]
      }
    end
  end

  def get_policy(dom,pol)
    (self.searchForPolicy :enterprise => {:domain => dom}, :policy => {:searchTerm => pol})['policies'].each do |policy|
      puts "#{policy['name']}"
      self.getPolicy :policy => {:id => policy['id']}
    end
  end

end

es = EnterpriseServer.new(:host => 'docomo-test-es.3lm.com')
#es.update_policy('test1.3lm.com','3LM API Test')

# es.registerAdmin :user => {:email => 'threelm@test1.3lm.com', :password => 'threelm'}
#es.searchForPolicy :enterprise => {:domain => 'test1.3lm.com'}
#es.removePolicy :policy => {:id => '10006:999'}

# (es.searchForUser :user => {:email => 'k@test1.3lm.com' })['users'].each do |user|
#   user['devices'].each do |dev|
#     es.pushCommandsToDevice :device => {:id => dev['id']}, :commands => [{:key => 'lockout', :value => '1111'}]
#   end
# end
es.print_device('k@test1.3lm.com')

File: api/apitest.py
#!/usr/local/bin/python

import urllib
import urllib2
import json
import sys
import base64

import httplib, mimetypes, io

BOUNDARY = '439763ed5b2bac8cc359964bc1efca424751d25a27efbf50ef3d24fdd1dd21ab'
CRLF = '\r\n'


DEBUG = 1
class AuthError(Exception):
    def __init__(self,value):
        self.value = value
    def __str__(self):
        return repr(self.value)

class EnterpriseServer:
    def __init__(self):
        try:
            self.host = 'docomo-test-es.3lm.com'
            self.port = 8443
            self.user = 'superadmin@docomo_mt.3lm.com'
            self.pswd = '1rr@5hA1M@53!'
            self.base_url = "https://%s:%s/api/mt/" % (self.host,self.port)
            self.token = ''
            resp = self.post('getSessionToken',{'username': self.user, 'password': self.pswd})
            self.token = resp['sessionToken']
        except KeyError:
            raise AuthError("Failed to login Enteprise Server")

    def urlopen(self,url,jd):
        data = urllib.urlencode({'json':json.dumps(jd)})
        if DEBUG == 1:
            print ">>> %s" % url
            print json.dumps(jd,indent=4)
        resp = json.loads(urllib2.urlopen(url ,data).read())
        if DEBUG == 1:
            print "<<<"
            print json.dumps(resp,indent = 4)
        return resp
        
    def post(self,cmd,jd):
        if self.token:
            jd['sessionToken'] = self.token
        url = self.base_url+cmd
        return self.urlopen(url,jd)

    def encode_uploadApk_formdata(self,domain,name,filename):
        token = self.post('getSessionToken',
                          {'username': "%s@%s" % (self.user,domain),
                           'password': self.pswd})['sessionToken']

        form = ('--' + BOUNDARY,
                'Content-Disposition: form-data; name="sessionId"',
                '',
                token,
                '--' + BOUNDARY,
                'Content-Disposition: form-data; name="apkName"',
                '',
                name,
                '--' + BOUNDARY,
                'Content-Disposition: form-data; name="appBinaryUploadElement"; filename="%s"' % filename, 
                'Content-Type: application/octet-stream',
                '',
                open(filename).read(),
                '--' + BOUNDARY + '--')

        ios = io.BytesIO()
        for line in form:
            ios.write(str(line))
            ios.write(CRLF)
        
        return ios.getvalue()

    def uploadApk_multipart(self,domain,name,filename):
        data = self.encode_uploadApk_formdata(domain,name,filename)

        h = httplib.HTTPSConnection(self.host,8443)
        h.putrequest('POST', '/adminconsole/appUploader')
        h.putheader('content-type', 'multipart/form-data; boundary=%s' % BOUNDARY)
        h.putheader('content-length', str(len(data)))
        h.endheaders()
        h.send(data)
        return h.getresponse().read()

    def uploadApk(self,domain,name,filename):
        data = open(filename,'r').read()
        data_base64 = base64.b64encode(data)
        self.post('uploadApk',
                  {'enterprise': {'domain': domain},
                   'apk': {'contents': data_base64, 'name' : name}})
         
es = EnterpriseServer()
print es.uploadApk_multipart('test1.3lm.com','Test Apk 011',sys.argv[1])




